internal/clients/team4/president.go

```diff
@@ -0,0 +1,351 @@
+package team4
+
+import (
+	"github.com/SOMAS2020/SOMAS2020/internal/common/baseclient"
+	"github.com/SOMAS2020/SOMAS2020/internal/common/config"
+	"math/rand"
+	"sort"
+
+	"github.com/SOMAS2020/SOMAS2020/internal/common/rules"
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+)
+
+type president struct {
+	*baseclient.BasePresident
+	parent *client
+	config.ClientConfig
+}
+
+type IslandRequestPair struct {
+	Key   shared.ClientID
+	Value shared.Resources
+}
+
+func rankByAllocationSize(allocationRequests map[shared.ClientID]shared.Resources) IslandRequestPairList {
+	pl := make(IslandRequestPairList, len(allocationRequests))
+	i := 0
+	for k, v := range allocationRequests {
+		pl[i] = IslandRequestPair{k, v}
+		i++
+	}
+	sort.Sort(sort.Reverse(pl))
+	return pl
+}
+
+// IslandRequestPairList implements sort.Interface for []IslandRequestPair
+type IslandRequestPairList []IslandRequestPair
+
+func (p IslandRequestPairList) Len() int           { return len(p) }
+func (p IslandRequestPairList) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+func (p IslandRequestPairList) Less(i, j int) bool { return p[i].Value < p[j].Value }
+
+// EvaluateAllocationRequests sets allowed resource allocation based on each islands requests
+func (p *president) EvaluateAllocationRequests(resourceRequest map[shared.ClientID]shared.Resources, availCommonPool shared.Resources) shared.PresidentReturnContent {
+	_, budgetRuleInPay := p.GameState.RulesInfo.CurrentRulesInPlay["president_over_budget"]
+	actionOverBudget := p.parent.ServerReadHandle.GetGameState().IIGORolesBudget[shared.President]-
+		p.parent.ServerReadHandle.GetGameConfig().IIGOClientConfig.ReplyAllocationRequestsActionCost < 0
```

Comment:
These should be using a function from utilities. If it's not there maybe just implement it? There should be a getBudget(shared.Roles).